### DISCUSSION TITLE
How to get plugins to load out before the YOURLS's language

### DISCUSSION THREAD

**Discussion by 8MiYile** [YOURLS/YOURLS#3677] [Plugins]
https://github.com/YOURLS/YOURLS/issues/3674

Because I don't know how to get my plugin to show the page before it loads up.
The solution I'm using is to have the plugin HOOK the $yourls_locate variable parameter
But there is no way to make the whole site all other languages, only plug-ins to load the language, did not achieve the effect of global application
I was hoping someone could give me some pointers on how to fix this.

Here's the source code for my plugin, which I haven't posted here because it's not yet stabilized.
https://github.com/8Mi-Tech/yourls-i18n-manager
```php
yourls_add_action( 'plugins_loaded', 'i18n_manager_plugin_loaded' );
function i18n_manager_plugin_loaded() {
    i18n_manager_set_language();
    yourls_load_custom_textdomain( 'i18n_manager', dirname( __FILE__ ) . '/languages' ); // 加载自身语言包
    if ( $_SERVER[ 'REQUEST_METHOD' ] === 'POST' && $_SERVER["QUERY_STRING"] === 'page=i18n-manager' ) { // 检查请求方法是否为 POST
        i18n_manager_process_request(); // 如果是 POST 请求，执行插件逻辑
    } else {
        yourls_register_plugin_page( 'i18n-manager', yourls__( 'I18N Manager' ,'i18n_manager'), 'i18n_manager_html' ); // 如果不是 POST 请求，注册设置页面
    }
}

i18n_manager_set_language();
function i18n_manager_set_language(){
    global $yourls_locale;
    $browserLang = current(explode(',', isset($_SERVER['HTTP_ACCEPT_LANGUAGE']) ? $_SERVER['HTTP_ACCEPT_LANGUAGE'] : ''));
    $json_languages = json_decode( file_get_contents( dirname( __FILE__ ).'/languages.json' ), true );
    foreach ($json_languages as $json_language) {
        if (in_array($browserLang, $json_language['aliases'])) {
            $matchedCode = $json_language['code'];
            break;
        }
    }
    $langFilePath = YOURLS_ABSPATH . '/user/languages/' . $matchedCode . '.mo';
    if (file_exists($langFilePath) && is_readable($langFilePath) && (!is_link($langFilePath) || (is_readable(readlink($langFilePath)) && filesize($langFilePath) > 60))) {                
        // TODO: Here, logic for checking the ".mo" language file needs to be inserted.
        $yourls_locale = is_link( $langFilePath ) ? pathinfo( readlink( $langFilePath ) )[ 'filename' ] : $matchedCode;
    }
    if (!isset($yourls_locale) or $yourls_locale === '' ) {
        $yourls_locale = 'en_US';
    }
    return $yourls_locale;
}
}
```
![image](https://github.com/YOURLS/YOURLS/assets/25455400/4df16aa1-03fa-427f-b6ba-a7da526d360d)

I've tried many things, such as changing the way it loads, and it still doesn't apply to the whole site, like in the image above
```php
yourls_add_action( 'activated_plugin', 'i18n_manager_set_language' );
// yourls_add_action( 'plugins_loaded', 'i18n_manager_set_language' );
```

**Comment by ozh** [comment]
See pull request #3679

**Comment by ozh** [comment]
Also, @8MiYile , with kind respect : it's sometimes hard to understand what you write. We get that not everybody writes perfect English, and for most of the core YOURLS teams English is not even their mother tongue. To help everyone understand what everybody writes, it's best to keep messages short and straight to the point of one specific issue.

All your posts so far have been super long, super detailed with code snippets and stuff, but mostly have failed to clearly address the issue you are trying to tackle. If I understand correctly, all your previous posts so far can be summarized with : "I want a plugin to modify the YOURLS_LANG setting, how to do this ?"

Posting code of what does not work as expected but failing to clearly express your initial need makes it difficult to understand your initial need.

**Comment by 8MiYile** [comment]
@ozh 
In fact, I am afraid you can not understand, and then I am Chinese to English,
I should be the one to say sorry, because I did not know that your main language of speech is not English

In a nutshell, how can plug-ins take over the original language system

And how do I write and which specific hook to connect to

The very long code area is mainly to tell you how I wrote this plugin, and to share it so that you can repeat this error
